test/hotspot/jtreg/serviceability/sa/ClhsdbDumpheap.java

```diff
@@ -51,26 +61,133 @@ public static void printStackTraces(String file) {
         }
     }
 
+    private static void verifyDumpFile(File dump, boolean compression) throws Exception {
+        assertTrue(dump.exists() && dump.isFile(), "Could not create dump file " + dump.getAbsolutePath());
+        if (!compression) {
+            printStackTraces(dump.getAbsolutePath());
+            return;
+        } else {
+            String deCompressedFile = "SAdump" + System.currentTimeMillis() + ".hprof";
+            File out = new File(deCompressedFile);
+            try {
+                GZIPInputStream gis = new GZIPInputStream(new FileInputStream(dump));
```

Comment:
`printStackTraces()` uses `Reader.getStack()`, which does not know about gzipped hprof files. However, `Reader.readFile()` was modified to automatically detect gzipped hprof files. I suggest you do the same for `getStack()` rather than making the test do all the work. Probably `getStack()` and `readFile()` can share the code that does this.

Reply:
Hi Chris, 
Thanks for review and nice catch on this.

I have considered using Reader.readFile() but it can not parse the gziped heap dump successfully, and I investigated the reason is that the underlying GzipAccess class requires the gziped heap dump file to have "HPROF BLOCKSIZE=xxx" bytes in the gziped file header.
I also investigated that the implementation of jmap command in jdk.jcmd added these words as a "comment section" in gz file headers, by passing it to the "ZIP_GZip_Fully()" when it is first called. 
But I cannot find a way to do same thing in GZIPOutputStream(), it's writeHeader() does not support "comment" section. 
Also I have verified the gziped heap dump file generated with GZIPOutputStream() could be parsed by heaphero.io successfully. 

So it seems to me that "HPROF BLOCKSIZE" may not be a must for general gziped heap dump, but I am not sure whether it is required for testing the jmap command, so I didn't touch GzipAccess in test and do decompress manually in ClhsdbDumpheap.java. 

I am considering another way to modify GzipAccess for testing to accept heap dump file without "HPROF BLOCKSIZE" header, I will try to investigate whether it could work. 
What do you think? 

Thanks!
Lin

Reply:
Sorry for typo. the GzipAccess should be GzipRandomAccess.

Reply:
So the problem is that the code in GzipRandomAccess.getAccess() returns NULL if it doesn't find the "HPROF BLOCKSIZE" comment, but you are also saying that tools like heaphero.io work fine without the comment. So maybe you just need to fix GzipRandomAccess.getAccess() to use a default blocksize if none is specified.

On the other hand, it seems it would be best if the SA hprof file was no different than the jcmd hprof file, and the SA file also included the comment. But as you said GZIPOutputStream() does not support adding a comment in the same manner as the zlib library does. I'm not sure how you would go about adding the support, so probably it is not worth it.

Reply:
As for any testing requirement, maybe you could pass in a flag to Reader indicating whether or not the testing requires the "HPROF BLOCKSIZE"  comment.

Reply:
Hi Chris，
I did more investigation today, IMHO, it is too complicated to modify the GzipRandomAccess because there are lots codes required to be modified in getAccess() and read() to handle the differences of headers, and also it may not be ok to define a default value for "HPROF BLOCKSIZE=" because the value seems to be used as a hint for decompression. 
On the other hand, I think it is possible to define a private class in HeapHprofBinWriter.cpp  that extends GZIPOutputStream, and then overwrite writeHeader() to add HPROF BLOCKSIZE as file commends, I saw there is logic in GZipInputStream's readHeader() that process this kind of info as file comments.  Then we can have same format of gziped heap dump file from jmap. 
I am working on the patch and will update when it is ready.

Thanks.
Lin

Reply:
Dear @plummercj,
I have made  investigation on extending GZIPOutputStream,  since it is not possible to override/overwrite private method writeHeader(), the only way I could figure out is to create a class named HProfGZIPOutputStream that extends DeflaterOutputStream,  copies most of GZIPOutputStream's code and rewrite a writeHeader() file. 

I am not sure whether it is nice since most of the code are same with GZIPOutputStream.  IMO, it may be nice if we can add another writeHeader() method to GZIPOutputStream that support adding file comments, but I am also not sure whether it is acceptable to add a new method in this general used class. 

BRs,
Lin

Reply:
I don't think we want to essentially clone GZIPOutputStream into a new HProfGZIPOutputStream in order to solve this problem. I think adding a GZIPOutputStream.writeHeader() method that supports adding a comment make sense, but probably not something that should be done with this CR. I new CR can be filed for it, and maybe take care of it sometime in the future, but for this we should focus on a solution that doesn't require it and isn't too complicated. 

Looking back at my original suggestion to put the gzip reading support in Reader.getStack(), I'm not clear on the issue you ran into. You said that GzipRandomAccess requires the "HPROF BLOCKSIZE=" comment, but I'm not so sure you need to use GzipRandomAccess. You can use the same gzip detection mechanism as Reader.readFile() does to see if it can read the file without unzipping it, and if not just do what your test currently does now, and try to read it with GZIPInputStream. You don't need to use GzipRandomAccess here.

Reply:
Thanks Chris! 
I think I digressed from the case. Thanks for point it out and pull me back. Your solution sounds great, I will try to make the change.

Reply:
Dear Chris,
Thanks for your suggestion, I digressed from the test case and focus too much on the GzipRandomAccess. Thanks for pull me back.
And it is hard for me to read file without unzipping it, so I made a change to add unzipping in Reader.getStack(). 
Would you like to help reveiw again?

BRs,
Lin